vim-patch:9.1.{0147,0149,0150,0152,0156},58f1e5c0893a: 'winfixbuf'

ColinKennedy opened the pull request:
Close #12517

#### vim-patch:9.1.0147: Cannot keep a buffer focused in a window

Problem:  Cannot keep a buffer focused in a window
          (Amit Levy)
Solution: Add the 'winfixbuf' window-local option
          (Colin Kennedy)

closes: vim/vim#13903

https://github.com/vim/vim/commit/215703563757a4464907ead6fb9edaeb7f430bea

Co-authored-by: Colin Kennedy <colinvfx@gmail.com>


#### vim-patch:9.1.0149: null pointer member access when accessing 'winfixbuf' property

Problem:  qf_goto_win_with_qfl_file may check if prevwin has 'winfixbuf'
          set without checking if it's valid first.
Solution: Reverse the condition. Add a test, a modeline, and a missing
          CheckFeature. (Searn Dewar)

closes: vim/vim#14140

https://github.com/vim/vim/commit/5131f224da93f2e042a4b22545ef62b1b2ab8460

Co-authored-by: Sean Dewar <6256228+seandewar@users.noreply.github.com>


#### vim-patch:58f1e5c0893a

runtime(doc): Highlight the error message at *E1513*

https://github.com/vim/vim/commit/58f1e5c0893aed47d762f5698a1ebebc05ec9819

Co-authored-by: Christian Brabandt <cb@256bit.org>


#### vim-patch:9.1.0150: Several minor 'winfixbuf' issues

Problem:  several minor 'winfixbuf' issues exist, mostly relating to the
	  quickfix list
Solution: address them and adjust tests. Retab and reflow a few things too.
	  (Sean Dewar)

Things touched include:

- Replace the semsgs with gettext'd emsgs.

- Handle window switching in ex_listdo properly, so curbuf and curwin
  are kept in-sync and trigger autocommands; handle those properly.

- Don't change the list entry index in qf_jump_edit_buffer if we fail
  due to 'wfb' (achieved by returning FAIL; QF_ABORT should only be used
  if the list was changed).

- Make qf_jump_edit_buffer actually switch to prevwin when using `:cXX`
  commands **outside** of the list window if 'wfb' is set in curwin.
  Handle autocommands properly in case they mess with the list.

  NOTE: previously, it seemed to split if 'wfb' was set, but do nothing
  and fail if prevwin is *valid*. This behaviour seemed strange, and maybe
  unintentional? Now it aligns more with what's described for the `:cXX`
  commands in the original PR description when used outside a list window,
  I think.

- In both functions, only consider prevwin if 'wfb' isn't set for it;
  fallback to splitting otherwise.

- Use win_split to split. Not sure if there was a specific reason for
  using ex_splitview. win_split is simpler and respects modifiers like
  :vertical that may have been used. Plus, its return value can be checked
  for setting opened_window in qf code (technically win_split_ins autocmds
  could immediately close it or change windows, in which the qf code might
  close some other window on failure; it's already the case elsewhere,
  though).

closes: vim/vim#14142

https://github.com/vim/vim/commit/4bb505e28cac0389561fff78d8bbe0319c2bcf2f

Co-authored-by: Sean Dewar <6256228+seandewar@users.noreply.github.com>

ColinKennedy committed:
vim-patch:9.1.0147: Cannot keep a buffer focused in a window

Problem:  Cannot keep a buffer focused in a window
          (Amit Levy)
Solution: Add the 'winfixbuf' window-local option
          (Colin Kennedy)

fixes:  vim/vim#6445
closes: vim/vim#13903

https://github.com/vim/vim/commit/215703563757a4464907ead6fb9edaeb7f430bea

N/A patch:
vim-patch:58f1e5c0893a

ColinKennedy committed:
vim-patch:9.1.0149: null pointer member access when accessing 'winfixbuf' property

Problem:  qf_goto_win_with_qfl_file may check if prevwin has 'winfixbuf'
          set without checking if it's valid first.
Solution: Reverse the condition. Add a test, a modeline, and a missing
          CheckFeature. (Searn Dewar)

closes: vim/vim#14140

https://github.com/vim/vim/commit/5131f224da93f2e042a4b22545ef62b1b2ab8460

Co-authored-by: Sean Dewar <6256228+seandewar@users.noreply.github.com>

zeertzjq commented:
Hmm, `has('pythonx')` is always true, which seems like a bug. I'll open a PR.

ColinKennedy commented:
> Hmm, has('pythonx') is always true, which seems like a bug. I'll open a PR.

Yeah I was wondering about that one but figured it was just an intended difference between Neovim and Vim. Glad you caught it :)

ColinKennedy commented:
I've updated the PR, please let me know if there's any other changes needed and I can add them. Thank you!

ColinKennedy commented:
I recently found out about https://github.com/vim/vim/pull/14142 and https://github.com/vim/vim/commit/4bb505e28cac0389561fff78d8bbe0319c2bcf2f. I'll add this patch into this PR and update the title/description

ColinKennedy committed:
vim-patch:9.1.0150: Several minor 'winfixbuf' issues

Problem:  several minor 'winfixbuf' issues exist, mostly relating to the
          quickfix list
Solution: address them and adjust tests. Retab and reflow a few things too.
          (Sean Dewar)

Things touched include:

- Replace the semsgs with gettext'd emsgs.

- Handle window switching in ex_listdo properly, so curbuf and curwin
  are kept in-sync and trigger autocommands; handle those properly.

- Don't change the list entry index in qf_jump_edit_buffer if we fail
  due to 'wfb' (achieved by returning FAIL; QF_ABORT should only be used
  if the list was changed).

- Make qf_jump_edit_buffer actually switch to prevwin when using `:cXX`
  commands **outside** of the list window if 'wfb' is set in curwin.
  Handle autocommands properly in case they mess with the list.

  NOTE: previously, it seemed to split if 'wfb' was set, but do nothing
  and fail if prevwin is *valid*. This behaviour seemed strange, and maybe
  unintentional? Now it aligns more with what's described for the `:cXX`
  commands in the original PR description when used outside a list window,
  I think.

- In both functions, only consider prevwin if 'wfb' isn't set for it;
  fallback to splitting otherwise.

- Use win_split to split. Not sure if there was a specific reason for
  using ex_splitview. win_split is simpler and respects modifiers like
  :vertical that may have been used. Plus, its return value can be checked
  for setting opened_window in qf code (technically win_split_ins autocmds
  could immediately close it or change windows, in which the qf code might
  close some other window on failure; it's already the case elsewhere,
  though).

closes: vim/vim#14142

https://github.com/vim/vim/commit/4bb505e28cac0389561fff78d8bbe0319c2bcf2f

Co-authored-by: Sean Dewar <6256228+seandewar@users.noreply.github.com>

clason commented on .github/workflows/docs.yml:
```diff
@@ -1,5 +1,8 @@
 name: docs
 on:
+  push:
+    branches:
+      - 'vim-9.1.0147-002'
```
👀 

(Convention is that such changes go in a separate `TEST COMMIT DELETE LATER` commit, not squashed in with the actual changes.)

ColinKennedy replied:
Crap I forgot to remove that, sorry. I'll try to remember in the future or at least keep that commit separate, as you suggested

clason replied:
No worries; that _is_ something that people do if they need to verify something specific on CI.

ColinKennedy commented on test/functional/options/winfixbuf_spec.lua:
```diff
@@ -0,0 +1,54 @@
+local clear = helpers.clear
+local exec_lua = helpers.exec_lua
+
+describe("Neovim API calls with 'winfixbuf'", function()
```
It's pretty late where I live so I will leave that reading for another day but for now I'll commit your suggestion!

ColinKennedy commented:
Okay, this PR has all patches for `'winfixbuf'` that I could see using `vim-patch.sh`. It should be reviewable tests pass (at the time of writing, a couple are still running). Please let me know of any edits that I should make!

zeertzjq committed:
vim-patch:9.1.0152: Coverity complains about ignoring return value

Problem:  Coverity complains about ignoring return value of win_split()
          (after v9.1.150)
Solution: Check if win_split() failed, add winfixbuf.res to Makefile

https://github.com/vim/vim/commit/af7ae8160041e2d17c56945381e9370e7178e596

Co-authored-by: Christian Brabandt <cb@256bit.org>

zeertzjq committed:
vim-patch:9.1.0156: Make 'wfb' failing to split still report E1513

Problem:  may not be clear why failing to split causes an ":Xdo" command
          to abort if 'wfb' is set.
Solution: do not return immediately if win_split fails, so E1513 is
          still given. Expect both errors in the test. Also fix tests to
          pass CI.
          (Sean Dewar)

closes: vim/vim#14152

https://github.com/vim/vim/commit/769eb2d0c3614f9ea6fffa82329558f1a4af384f

Co-authored-by: Sean Dewar <6256228+seandewar@users.noreply.github.com>
